### PR TITLE
Add TypeScript proxy to support JrS debugging w/ Chrome DevTools

### PR DESCRIPTION
**Pull request by grgustaf**
Basic debugging features of step over, step in , step out, continue, stop,
add/remove breakpoints, provide backtrace at breakpoint, and evaluate watch
expressions are all working.

A few low-hanging features still missing:
- passthrough of output to CDT console
- reporting exceptions
- verbosity control

Other features of CDT will require enhancement to JrS debugger, e.g.:
- scope chain
- inspectable eval output (e.g. JSON, not just strings)
- memory analysis
- profiling

Significant level of unit testing provided via jest.

Thanks to Martijn The <martijn.the@intel.com> for initial TS code skeleton,
some unit tests, 'step out' feature, and tons of review!

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

### CODE REVIEW

**Comment by grgustaf** [comment]
@martijnthe you could check my attempt to merge the `package.json` and `README.md` to make sure they make sense.

Also, Travis failed I think because we haven't done `yarn install`. Not sure off-hand where the right place to put that is.

**Comment by martijnthe** [comment]
> Also, Travis failed I think because we haven't done yarn install. Not sure off-hand where the right place to put that is.

I had added `yarn` which is the shorthand for `yarn install`.
The `yarn.lock` file is missing, that's why it's not working, I think.

**Comment by martijnthe** [comment]
@knightburton we've moved over the PR from https://github.com/jerryscript-project/jerryscript/pull/2205 to here. Can you take another review pass? I've asked @akosthekiss to add you to the team. Thanks.

**Comment by martijnthe** [comment]
@grgustaf actually, there's more missing:

- `tsconfig.json`
- `tslint.json`
- `yarn.lock`
- `jerry-debugger.sh`

**Comment by grgustaf** [comment]
Yep, missed those files - just updated with the additions.

**Comment by martijnthe** [comment]
LGTM

File: src/lib/debugger-client.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      this.socket.terminate();
+      this.socket = undefined;
+    }
+  }
+
+  onMessage(data: ArrayBuffer) {
+    this.delegate.onMessage(new Uint8Array(data));
+  }
+
+  send(data: ArrayBuffer) {
```
Comment (knightburton) [inline src/lib/debugger-client.ts]: Maybe I am wrong, however I think this `data` parameter should be `Uint8Array`.

**Review by knightburton — COMMENTED** [review]
@martijnthe I've ran through the changes again.

File: src/lib/debugger-client.ts (hunk 1 above)
Comment (martijnthe) [reply src/lib/debugger-client.ts]: Yeah, you're right, thanks for catching.
I'm a bit surprised the TypeScript compiler didn't throw an error about this.

@grgustaf will you fix it up?

**Comment by grgustaf** [comment]
Thanks yeah I'll be back tomorrow and will take care of that, sorry for the delay.

**Comment by grgustaf** [comment]
Fixed that line, thanks...

**Comment by martijnthe** [comment]
LGTM, thanks.

@knightburton OK now?

**Comment by knightburton** [comment]
@martijnthe @grgustaf Yes, all good! :) LGTM

**Comment by martijnthe** [comment]
@zherczeg @akosthekiss @yichoi Do you have an interest in reviewing this?
If not, we'll go ahead and merge this. Myself and @knightburton have reviewed it already and have +1'd it.

**Comment by akosthekiss** [comment]
@martijnthe I've skimmed through the patch, and as much as an expert I am with TS, the patch LGTM

**Comment by zherczeg** [comment]
You can setup your own reviewing policies for this project. The only requirement is apache 2.0 license and DCO for clear reusability.

**Comment by grgustaf** [comment]
OK, thanks...

**Pull request merged**